design-proposals/klip-15-new-api-and-client.md

```diff
@@ -0,0 +1,189 @@
+A further advantage of using Vert.x on the server is that Vert.x has support for GraphQL which will make the implementation of
+GraphQL for ksqlDB relatively simple compared to having to implement it from scratch.
+
+## Public APIS
+
+The following changes / additions to APIs will occur:
+
+* The current chunked streaming and websockets streaming endpoints will be retired.
+* The client will provide a new public API
+* We will provide a new wire protocol for client<->server communication.
+
+## Design
+
+### The client
+
+* The Java client will provide an API based on JDK 9 flow / reactive streams and completable future
+* The networking will be handled by Vert.x (which uses Netty). Both Vert.x and Netty are very widely
+used toolkits which are known for great performance, and have few dependencies.
+* The client will have minimal transitive jar dependencies - this is important as the client will be embedded in end user applications.
+* Client connections are designed to be re-used.
+* The client will be thread-safe.
+
+### The server
+
+* The toolkit used on the server side will be Vert.x
+* The current Jetty / JAX-RS usage will be retired.
+* The current non streaming HTTP/REST endpoints will be migrated to Vert.x - this should modernise and radically simplify and
+clarify the server side implementation result in a cleaner implementation and reduced lines of code.
+* The current query streaming endpoints (chunked response and Websockets) will be retired.
+* Any current Jetty specific plugins (e.g. security plugins) will be migrated to Vert.x
+* Vert.x has great support for working with various different network protocols and has has [unrivalled performance/scalability](https://www.techempower.com/benchmarks/)
+characteristics for a JVM toolkit. It will also set us up well for a fully async / reactive internal threading model in the server that we
+should aim towards for the future.
+
+### Implementation plan
+
+There is a lot of work to do here and it should be split up into separate work packages rather than trying to complete it as a single
+monolithic piece of work. Some obvious partitions of the work would be:
+
+* Protocol implementation
+* Java client implementation. This can be developed using a fake server so there is no dependency on completing the server work first.
+* JavaScript client implementation. Again, this can be developed using a fake server
+* Server streaming implementation
+* Migration of existing endpoints to new server implementation
+* Migration of existing plugins to new server implementation.
+* Switchover from old API to new API
+
+We should retain the current API implementation until it's ready to be switched over to the new implementation. The new implementation can exist
+in the same codebase during development, and only after switchover will the old implementation be removed.
+That way we can ensure a seamless switchover as there is a working API at all times.
+
+## Test plan
```

Comment:
We should have stress or performance test plan too so we can catch bugs when running several clients. Sometimes severs can hang due to the number of available threads it can handle.